### PR TITLE
P3 Autoconversion: Physics Property Tests and Technical Documentation

### PR DESCRIPTION
**Pull request by pressel**
P3 Autoconversion: Physics Property Tests and Technical Documentation

**Description:**
This PR introduces a new "Physics Property Test" strategy for the P3 microphysics scheme, starting with the cloud water to rain autoconversion process. Unlike traditional BFB regression tests which only detect changes from a baseline, these tests validate that the implementation adheres to the fundamental physical principles of the parameterization (Khairoutdinov and Kogan, 2000).

In addition, this PR establishes a new section in the Technical Guide for detailed physics documentation.

**Key Changes:**

1.  **Enhanced Unit Tests (`p3_autoconversion_unit_tests.cpp`)**:
    *   **Phase Space Sweep**: Tests cover a wide range of regimes ($q_c$ from $5 \times 10^{-9}$ to $10^{-2}$ kg/kg, $N_c$ from $10^6$ to $10^9$ #/m³).
    *   **Physical Properties Validated**:
        *   **Thresholds**: Verifies rate is zero below $q_c = 10^{-8}$ kg/kg.
        *   **Monotonicity**: Checks correct sensitivity to inputs ($\partial R/\partial q_c > 0$ and $\partial R/\partial N_c < 0$).
        *   **Conservation**: Verifies specific mass loss equals specific number loss ($N_{loss} = R \cdot N_c/q_c$).
        *   **Limits**: Verifies autoconversion is negligible in the "Haze Limit" ($r < 1\mu m$).
        *   **Self-Consistency**: Verifies the implicit mass of new raindrops corresponds to the characteristic radius ($r_{auto} \approx 25\mu m$).
    *   **Parameter Validation**: Ensures runtime configuration matches KK2000 constants.

2.  **New Documentation (`docs/technical/physics/p3/autoconversion.md`)**:
    *   Documents the mathematical formulation of the autoconversion rate and number tendencies.
    *   Explains the implementation details (thresholds, variance scaling status).
    *   Details the test strategy and the specific physical properties being verified.

3.  **Documentation Structure**:
    *   Updated `components/eamxx/mkdocs.yml` to include a new **Physics** section under the **Technical Guide**.

**Testing:**
*   Ran `p3_autoconversion_unit_tests` on `pm-cpu` (GNU).
*   All property checks pass.
*   The test correctly identifies that subgrid variance scaling is currently disabled in the implementation.

### CODE REVIEW

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 1]
```diff
@@ -123,33 +124,360 @@ void cloud_water_autoconversion_unit_bfb_tests() {
...
+        // =====================================================================
+        
+        Real mean_mass = qc / nc;
+        // Correct formula: r = (3*m / (4*pi*rho))^(1/3)
+        // rho_w = 1000.
+        Real mean_rad = std::pow((3.0 * mean_mass) / (4.0 * 3.14159 * 1000.0), 1.0/3.0);
+        
+        // A. Haze Limit
+        if (mean_rad < 1e-6) {
+            if (R > 1e-15) {
```
[the remark this answers is not in the document]
Comment (pressel) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp]: Actually, a threshold of 1e-25 would be too strict. Analytic calculation shows that for large haze droplets ($r \approx 0.9 \mu m$, $N_c=10^9 m^{-3}$), the KK2000 rate is $\approx 1.4 \times 10^{-16} s^{-1}$. The current threshold of 1e-15 correctly classifies this as negligible. Lowering it to 1e-20 or 1e-25 would cause false failures for physically valid (but negligible) rates.

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 2]
```diff
@@ -123,33 +124,372 @@ void cloud_water_autoconversion_unit_bfb_tests() {
...
+            Real ratio = h_rate_2(i) / h_rate_1(i);
+
+            // Detection: Do standard and variance rates differ?
+            if (std::abs(ratio - 1.0) > detect_eps) {
+                is_variance_scaling_active = true;
+                
+                // Validation: If active, is the rate suppressed? (Physical violation)
+                // Jensen's inequality for convex functions implies enhancement (ratio > 1).
+                // We fail if ratio < 1.0 (minus tolerance).
+                if (ratio < 1.0 - validate_eps) {
```
[the remark this answers is not in the document]
Comment (pressel) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp]: @copilot Won't fix this. The code is logically sound. Changing validate_eps creates commit noise without changing the behavior of the test. The current implementation safely distinguishes between "feature inactive" (ratio $\approx$ 1.0) and "feature broken" (ratio < 1.0).

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 3]
```diff
@@ -123,33 +124,372 @@ void cloud_water_autoconversion_unit_bfb_tests() {
...
+             // Expected: 25 um radius drop.
+             // m = 4/3 * pi * rho_w * r^3
+             // m(20um) ~ 3.35e-11 kg
+             // m(30um) ~ 1.13e-10 kg
+             const Real min_allowed_mass = 3.3e-11; 
+             const Real max_allowed_mass = 1.2e-10;
+
+             if (mass_drop < min_allowed_mass || mass_drop > max_allowed_mass) { 
+                 std::cout << "Embryo Size Fail: Mass=" << mass_drop 
+                           << " kg. Expected range [3.3e-11, 1.2e-10] (approx 20-30 um radius)\n";
```
[the remark this answers is not in the document]
Comment (pressel) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp]: fixed

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 4]
```diff
@@ -123,33 +124,372 @@ void cloud_water_autoconversion_unit_bfb_tests() {
...
+        // rho_w = 1000.
+        using C = scream::physics::Constants<Real>;
+        Real mean_rad = std::pow((3.0 * mean_mass) / (4.0 * C::Pi * 1000.0), 1.0/3.0);
+        
+        // A. Haze Limit
+        if (mean_rad < 1e-6) {
+            if (R > 1e-15) {
+                std::cout << "Haze Limit Fail: r=" << mean_rad << ", R=" << R << "\n";
+                failures++;
+            }
```
[the remark this answers is not in the document]
Comment (pressel) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp]: @copilot  I attempted to tighten this threshold to 1e-30 to match the rest of the file, but it caused the test to fail.

Because the autoconversion parameterization is a continuous power law (active whenever $q_c > 10^{-8}$), it yields extremely small but finite rates (e.g., $\sim 10^{-25}$) in the haze regime. While these rates are physically negligible compared to actual rain formation, they are not numerically zero.

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 5]
```diff
@@ -123,33 +124,372 @@ void cloud_water_autoconversion_unit_bfb_tests() {
...
+            // A. Colloidal Stability (Inverse Nc Dependency): dR/dNc < 0
+            // R_next (high Nc) should be less than R (low Nc).
+            if (R > absolute_floor) {
+                // If R_next > R, it's a failure. Allow small tolerance.
+                Real max_allowed_R_next = R * (1.0 + relative_tolerance);
+                if (R_next > max_allowed_R_next) {
+                     std::cout << "Monotonicity Nc Fail: R increased with Nc. qc=" << qc 
+                               << " R=" << R << " R_next=" << R_next << "\n";
+                     failures++;
+                }
```
[the remark this answers is not in the document]
Comment (pressel) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp:374]: I have updated the test to enforce strict monotonicity ($\partial R / \partial N_c < 0$). The code now fails if R_next >= R, ensuring that splitting the same water mass into more droplets always reduces the conversion rate, as physically required. I also updated the technical documentation to reflect this stricter check.

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 6]
```diff
@@ -123,33 +124,372 @@ void cloud_water_autoconversion_unit_bfb_tests() {
...
+    int failures = 0;
+
+    for (int j = 0; j < n_nc; ++j) {
+      for (int i = 0; i < n_qc; ++i) {
+        int idx = j * n_qc + i;
+        Real qc = qc_host(idx);
+        Real nc = nc_host(idx);
+        Real R = qc2qr_host(idx);
+        Real N_loss = nc2nr_host(idx);
+        Real R_ncautr = ncautr_host(idx);
```
[the remark this answers is not in the document]
Comment (pressel) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp]: no need to change

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 7]
```diff
@@ -123,33 +124,409 @@ void cloud_water_autoconversion_unit_bfb_tests() {
...
+    // =========================================================================
+    // Test Tolerance Definitions
+    // =========================================================================
+    // Identity tolerances: precision-dependent (mathematical exactness)
+    // These verify that the implementation follows its documented formulas.
+#ifdef SCREAM_DOUBLE_PRECISION
+    const Real identity_tol = 1e-14;  // ~100x machine epsilon (double)
+#else
+    const Real identity_tol = 1e-7;   // ~1000x machine epsilon (float)
+#endif
```
Comment (mahf708) [inline components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp]: can also use something from std/kokkos numerics here for the automatically detected epsilon with `<Real>`

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp [hunk 8]
```diff
@@ -123,33 +124,409 @@ void cloud_water_autoconversion_unit_bfb_tests() {
     cloud_water_autoconversion_unit_bfb_tests();
   }
 
-  KOKKOS_FUNCTION static void autoconversion_is_positive(const Int &i, Int &errors){
-
-    const Spack rho(1.0), inv_qc_relvar(1.0);
-    Spack qc_incld, nc_incld(1e7), qc2qr_autoconv_tend(0.0), nc2nr_autoconv_tend(0.0), ncautr(0.0);
-    for(int si=0; si<Spack::n; ++si){
-        qc_incld[si] = 1e-6 * i * Spack::n + si;
+  void validate_autoconversion_parameters() {
```
Comment (mahf708) [inline components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp:124]: note that we set the defaults in the xmls and the hardcoded items below may cause fails that will require code edits here again? nbd, we can always edit the items below, just noting it here for your consideration

File: components/eamxx/docs/technical/physics/p3/autoconversion.md [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+mixing ratio $N_c$ ($\#/\text{kg}$) and air density $\rho$ ($\text{kg}/\text{m}^3$),
+where $N_c$ corresponds to the variable `nc_incld`:
+
+$$
+N_{c,\text{vol}} = N_c \cdot \rho \cdot 10^{-6}
+\;\equiv\; \texttt{nc\_incld} \cdot \texttt{sp(1.e-6)} \cdot \rho
+$$
+
+\equiv \texttt{nc_incld} \cdot \texttt{sp(1.e-6)} \cdot \rho
+### Number Tendencies
```
Comment (mahf708) [inline components/eamxx/docs/technical/physics/p3/autoconversion.md:46]: need a line above this "title" otherwise linter will keep crying

**Review by mahf708 — APPROVED** [review]
I skimmed this PR, and this looks pretty good. I added some minor comments which can be addressed later. Some equations in https://docs.e3sm.org/E3SM/pr-preview/pr-7999/EAMxx/technical/physics/p3/autoconversion/ aren't loading correctly in my browser, just double check that one more time before merging

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp (hunk 7 above)
Comment (mahf708) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp]: e.g.
```cpp
// A numerical tolerance
auto tol = std::numeric_limits<Real>::epsilon() * 100;
```

**Comment by mahf708** [comment]
Also, please consider a rebase to reduce the number of commits from 25 to a meaningful few: `git rebase -i HEAD~25`

**Comment by rljacob** [comment]
Waiting on @pressel to fix some things.

**Review by mahf708 — CHANGES_REQUESTED** [review]
Looks good, but let's fix the linter and let's rebase and squash into a commit or two; @rljacob and @bartgol aren't fans of "merge master into branch" commits (of which you have three out of 27) because they distort the git history

File: components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp (hunk 8 above)
Comment (bartgol) [reply components/eamxx/src/physics/p3/tests/p3_autoconversion_unit_tests.cpp:124]: unit tests know nothing about xml files though...

**Comment by bartgol** [comment]
I don't know much about p3, so my comments would prob not help much. Taking myself off the reviewers list.

One general comment on the physics testing infrastructure though: I think it would be nice if we could take advantage of Catch system of partitioning tests via TEST_CASE and (possibly nested) SECTION clauses. I would consider breaking `run_phsyics` into individual property test names, and wrap them inside a SECTION. Something like

```cpp
TEST_CASE("p3_cloud_water_autoconversion_test")
{
  using T = scream::p3::unit_test::UnitWrap::UnitTest<scream::DefaultDevice>::TestP3CloudWaterAutoconversion;

  T t;
  
  SECTION ("baselines") {
    r.run_bfb();
  }
  
  SECTION ("property_tests") {
     SECTION ("property1") {
       t.run_property1_check();
     }
     SECTION ("property2") {
       r.run_property2_check();
     }
  }
}
```
You could even put them all in separate TEST_CASES (easier to run them individually). The choice between 2 `TEST_CASE`'s vs 1 `TEST_CASE` with 2 `SECTION`'s inside boils down to how much code you need to "setup" the tests. Since in this case there's really no setup (you just default build the class), then multiple test cases are ok. If you need to load some stuff, and validade params, then maybe you can write that once in a single TEST_CASE before the SECTION clauses.

But this is just choreography, and only matters when one has to manually run tests to debug.

**Comment by pressel** [comment]
@bartgol Thanks for the helpful suggestion on test choreography. I’ve updated the property testing accordingly.

**Review by mahf708 — APPROVED** [review]

**Comment by mahf708** [comment]
Thanks @pressel, looks great!

**Comment by AaronDonahue** [comment]
I don't see any reason why the one SMS test would fail because of changes in this PR.  It fails during the build, but other CIME cases do not.  I'm going to rerun the test and see if it was a fluke.

**Comment by tcclevenger** [comment]
@AaronDonahue I agree, it must be unrelated.. We can test on master if it persists here

**Comment by tcclevenger** [comment]
Triggered the error on master, I'll merge the PR and investigate separately

**Comment by rljacob** [comment]
For future reference, make sure the PR description doesn't have any fancy formatting like superscripts or subscripts.  They don't render well in "git log --first-parent"

**Comment by AaronDonahue** [comment]
@pressel ^^ See Rob's comment above.
